Infinite loop in sqlquerycount for MySQL #3185

bsergean opened the pull request:
This relates to #3185, which is not MySQL related.

Using a re-entrant Lock is behaving better than using a standard Lock so far (after a couple hours of uptime).

https://docs.python.org/2/library/threading.html#rlock-objects
http://stackoverflow.com/questions/16567958/when-and-how-to-use-pythons-rlock

bsergean commented:
Ping @mitsuhiko + @mattrobenolt

mitsuhiko commented:
The rlock would only solve a hang caused by a deadlock here. However this means that we have a recursive call to the patcher which currently is not guarded against. I'm all for adding checks to it.

bsergean commented:
Alright, so if the rlock fix will hide the real problem, how should we proceed to fix it ? 
How can we find call sites of the patcher and protect them (with another lock ?).

I have a very long log file that contains quite a few of those errors which I can send out your way.

Below are all the call sites for the PatchContext in the master branch as of today, but I have to say that my understanding is that the call sites will be in third party libraries (like Django), so I don't understand how we could control something from there without patching third party code.
 
```
sentry/src/sentry/debug/panels/base.py
25:        >>>         PatchContext('foo.bar', FunctionWrapper(collector))

sentry/src/sentry/debug/panels/redis.py
9:from ..utils.patch_context import PatchContext
84:            PatchContext('redis.client.StrictRedis.execute_command', RedisWrapper(collector)),
85:            PatchContext('redis.client.BasePipeline.execute', RedisPipelineWrapper(collector)),

sentry/src/sentry/debug/utils/patch_context.py
7:class PatchContext(object):

sentry/src/sentry/utils/performance/sqlquerycount.py
8:from sentry.debug.utils.patch_context import PatchContext
74:        self._patcher = PatchContext('django.db.backends.BaseDatabaseWrapper.cursor', self._cursor)
```

ps: I haven't hit any of those original errors over 3 days of running the code with the rlock fix, for what it's worth.

bsergean commented:
I have reverted the RLock fix in my local instance, and I'm not using the regular Lock. I'll wait to see if I encounter the bug again.

I had left out some debugging code, and maybe this was the cause of my problem. I've removed this debugging code.

dcramer commented:
I'm fine with reverting the perf tools hooks for now as well. 






On Mon, May 16, 2016 at 12:32 PM -0700, "Benjamin Sergeant" <notifications@github.com> wrote:












I have reverted the RLock fix in my local instance, and I'm not using the regular Lock. I'll wait to see if I encounter the bug again.



I had left out some debugging code, and maybe this was the cause of my problem. I've removed this debugging code.



—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub

bsergean commented:
I don't mind one way or another. I'll wait a bit and if the error is gone after a day or so, it means that the Lock fix is sufficient (and good to release in a dot/patch release ;) ?).

mitsuhiko commented:
Just to confirm: RLock works for you but Lock itself does not?

bsergean commented:
Yes that's correct. But given the intermittent nature of this problem I don't know what we can make of it.

Here's a recollection of last week until today.

1. May 05 / first time I hit that problem, after upgrading to 8.4.0. My instance is pretty broken.
2. May 06 / trying to install locally (as in cp over that one file) the Lock fix. Works well for 3 days.
3. May 09 / the problem starts re-appearing
4. Friday 13 / put in the RLock fix, no crash since
5. rollback RLock fix to have only the Lock fix. Wait and see.

Here's a list of related errors, extracted from my journalctl log file.

```
 Error processing 'is_rate_limited' on 'RedisQuota': maximum recursion depth exceeded in __instancecheck__
 Error processing 'is_rate_limited' on 'RedisQuota': maximum recursion depth exceeded in __subclasscheck__
 Error processing 'is_rate_limited' on 'RedisQuota': maximum recursion depth exceeded in cmp
 Error processing 'is_rate_limited' on 'RedisQuota': maximum recursion depth exceeded while calling a Python object
 Error processing 'load_user_conf' on 'SentryLocaleMiddleware': maximum recursion depth exceeded
 Error processing 'load_user_conf' on 'SentryLocaleMiddleware': maximum recursion depth exceeded in cmp
 Error processing 'load_user_conf' on 'SentryLocaleMiddleware': maximum recursion depth exceeded while calling a Python object
 Not capturing event due to unsafe stacktrace:
 Task sentry.tasks.options.sync_options[3bdc3d6a-88a4-4274-9291-78d78dcf2e8d
 Unable to fetch ProjectKey for internal project
 Unable to fetch internal project for some unknown reason
 Unable to process log entry: maximum recursion depth exceeded while calling a Python object
 function takes exactly 1 argument (3 given)
 maximum recursion depth exceeded
 maximum recursion depth exceeded in __instancecheck__
 maximum recursion depth exceeded in __subclasscheck__
 maximum recursion depth exceeded in cmp
 maximum recursion depth exceeded while calling a Python object
```

Hope it helps.

mitsuhiko commented:
The reason I doubt that the RLock has a chance to fix something is that it's reentrant for the same thread. The difference between a regular lock is literally only that. Which means that you should not have seen a recursion depth problem but a complete deadlock as the thread would deadlock itself on that lock trying to patch something that is already patched.

bsergean commented:
Closing the PR. Indeed I did not see any deadlock when the problem seemed to "re-appeared"; what I did see was the infinite recursion loop.

Now that I think about it my previous memory of a C++ bug being fixed by using a recursive mutex was a hang which was likely to be a deadlock indeed.
